static/js/tools/choropleth/choropleth.tsx

```diff
@@ -0,0 +1,356 @@
+   */
+  loadGeoJson(): void {
+    let geoUrl = "/api/choropleth/geo";
+    geoUrl += buildChoroplethParams(["pc", "geoDcid", "level", "mdom"]);
+    let valueUrl = "/api/choropleth/values";
+    valueUrl += buildChoroplethParams(["geoDcid", "statVar"]);
+
+    // Create request and generate map.
+    const geoPromise = axios.get(geoUrl);
+    const valuePromise = axios.get(valueUrl);
+
+    Promise.all([geoPromise, valuePromise]).then((values) => {
+      this.setState({ geojson: values[0].data[0], values: values[1].data[0] });
+      //TODO(iancostello): Investigate if this can be moved to
+      //shouldComponentUpdate.
+      this.renderGeoMap();
+      this.updateGeoValues();
+    });
+  }
+
+  /**
+   * Loads values for the current geoDcid and updates map.
+   */
+  loadValues(): void {
+    let baseUrl = "/api/choropleth/values";
+    baseUrl += buildChoroplethParams(["geoDcid", "statVar"]);
+
+    axios.get(baseUrl).then((resp) => {
+      this.setState({ values: resp.data[0] });
+      this.updateGeoValues();
+    });
+  }
+
+  /**
+   * Handles rendering the basic blank geoJson map.
+   * Requires state geojson to be set with valid geoJson mapping object.
+   */
+  renderGeoMap(): void {
+    // Combine path elements from D3 content.
+    const geojson = this.state["geojson"];
+    const mapContent = d3
+      .select("#main-pane g.map")
+      .selectAll("path")
+      .data(geojson.features);
+
+    // Scale and center the map.
+    const svgContainer = document.getElementById("map_container");
+    const projection = d3
+      .geoAlbers()
+      .fitSize([svgContainer.clientWidth, svgContainer.clientHeight], geojson);
+    const geomap = d3.geoPath().projection(projection);
+
+    // Build map objects.
+    mapContent
+      .enter()
+      .append("path")
+      .attr("d", geomap)
+      // Add CSS class to each path for border outlining.
+      .attr("class", "border")
+      .attr("fill", "gray")
+      // Add various event handlers.
+      .on("mouseover", this.handleMapHover)
+      .on("mouseleave", this.mouseLeave)
+      .on("click", this.handleMapClick);
+
+    this.setState({ mapContent });
+
+    // Create population map.
+    const popMap = {};
+    for (const index in geojson.features) {
+      const content = geojson.features[index];
+      popMap[content.id] = content.properties.pop;
+    }
+    this.setState(() => {
+      return { popMap };
+    });
+  }
+
+  /**
+   * Updates geoJson map with current values loaded in state.
+   * Requires geoJson map to be rendered and state values to be set.
+   */
+  updateGeoValues(): void {
+    const values = this.state["values"];
+    const isPerCapita = this.state["pc"];
+
+    // Build chart display options.
+    const colorScale = d3
+      .scaleLinear()
+      .domain(
+        determineColorPalette(values, this.state["pc"], this.state["popMap"])
+      )
+      .range(["#deebf7", "#9ecae1", "#3182bd"] as any);
```

Comment:
try to use a tighter type than any

Reply:
This is a D3 library issue I believe, the type of range is explicitly set to on number[], but this isn't the case.